README.md
# SHA256 Check Extension for Nemo File Manager

**Current Version: 0.1.0**

This extension adds a right-click context menu item in the Nemo file manager on Linux Mint to compute and compare the SHA256 hash of files.

## Features

- Compute SHA256 hash for any selected file.
- Prompt for the expected hash value and compare it with the computed hash.
- Display the computed hash and indicate whether it matches the expected value.

## Installation

To install the SHA256 Check extension, follow these steps:

1. Ensure that `zenity` and `nemo` are installed on your system (these are typically installed by default on Linux Mint).
2. Download the latest release of SHA256 Check from the [GitHub repository](https://github.com/stephend9/sha256_check).
3. Extract the package to a known directory.
4. Open a terminal and navigate to the directory containing `install.sh`.
5. Run the installation script with the command `./install.sh`.

A progress window will appear showing the installation status. Once complete, a message will inform you that the installation has succeeded and Nemo will be restarted.

## Usage

After installation:

1. Right-click on any file in Nemo.
2. Select "Check SHA256 Hash" from the context menu.
3. Enter the expected SHA256 hash in the input box that appears.
4. A dialog will display the computed SHA256 hash and indicate whether the hashes match.

## Uninstallation

To uninstall the SHA256 Check extension, follow these steps:

1. Navigate to the directory containing `uninstall.sh`.
2. Run the uninstallation script with the command `./uninstall.sh`.

You will be asked to confirm the uninstallation. Please note that Nemo will restart as part of the uninstallation process, closing all open file browser windows.

## Dependencies

- Zenity
- Nemo File Manager

These dependencies are installed by default on Linux Mint.

## License

SHA256 Check is licensed under the GNU General Public License v3.0.

## Contributing

Contributions, issues, and feature requests are welcome. Feel free to check the [issues page](https://github.com/stephend9/sha256_check/issues) if you want to contribute.

## Contact

Stephen D - [@stephend9](https://twitter.com/stephend9) - stephend9@gmail.com

Project Link: [https://github.com/stephend9/sha256_check](https://github.com/stephend9/sha256_check)

## Donations

If you appreciate this extension and want to support its development, consider making a donation:

- **Bitcoin (BTC):** `bc1q3jyrlz7qmh2cfyglhxhpq62m2ky0eld2wx5xan`
- **Ethereum (ETH):** `0x246d84049Ad6763059643B09300F92aa4e5be7C8`

Your support is greatly appreciated!



CHANGELOG.md
# Changelog
All notable changes to SHA256 Check will be documented in this file.

## [0.1.0] - 2023-12-05
- Initial beta release.



uninstall.sh
#!/bin/bash

# Uninstall script for SHA256 Check - Version 0.1.0

# The directory where the script was installed
install_dir="$HOME/.local/bin"

# The directory where the Nemo action was installed
actions_dir="$HOME/.local/share/nemo/actions"

# Function to display a message using zenity
show_message() {
    zenity --info --text="$1" --width=300 --height=200
}

# Function to ask for confirmation
confirm() {
    zenity --question --text="$1" --width=300 --height=200
}

# Ask for confirmation before proceeding with uninstallation
if confirm "Are you sure you want to uninstall the SHA256 Check extension? This will remove all installed files and cannot be undone.\n\nPlease note: Nemo will restart as part of the uninstallation process, which will close all open file browser windows."; then

    # Remove the script and the .nemo_action file
    rm -f "$install_dir/sha256_check.sh"
    rm -f "$actions_dir/sha256_check.nemo_action"

    # Ask for confirmation before restarting Nemo
    if confirm "The uninstallation is complete. Nemo needs to restart, which will close all file browser windows. Do you want to continue?"; then
        # Restart Nemo
        nemo -q
        sleep 1
        nemo &

        # Wait for Nemo to restart
        sleep 2

        # Bring the Zenity window back to the front
        wmctrl -a "Uninstalling SHA256 Check"

        show_message "Nemo has been restarted. Uninstallation is complete."
    else
        show_message "Nemo has not been restarted. Please manually restart Nemo to complete the uninstallation."
    fi
else
    show_message "Uninstallation cancelled by the user."
fi



sha256_check.sh
#!/bin/bash
# SHA256 Check - Version 0.1.0

# Check if a file path is provided
if [ "$#" -ne 1 ]; then
    zenity --error --title="Error" --text="Usage: $0 <file_path>" --width=300
    exit 1
fi

# The file path is the first argument
file_path="$1"

# Check if the file exists
if [ ! -f "$file_path" ]; then
    zenity --error --title="Error" --text="Error: File does not exist." --width=300
    exit 1
fi

# Compute the SHA256 hash
computed_hash=$(sha256sum "$file_path" | awk '{ print $1 }')

# Prompt the user to enter the expected hash
expected_hash=$(zenity --entry --title="Verify SHA256 Hash" --text="Enter the expected SHA256 hash (max 64 characters):" --width=550 --entry-text "")

# If the user cancels the input box, exit
if [ -z "$expected_hash" ]; then
    exit 1
fi

# Check if the expected hash length is more than 64 characters
if [ "${#expected_hash}" -gt 64 ]; then
    zenity --error --title="Input Error" --text="The entered hash is too long. Maximum 64 characters allowed." --width=300
    exit 1
fi

# Compare the hashes and construct a message
if [ "$computed_hash" = "$expected_hash" ]; then
    match_message="✅ The SHA256 hash matches the expected value."
else
    match_message="❌ The SHA256 hash does not match the expected value."
fi

# Display the computed hash, expected hash, and the result of the comparison
zenity --info --title="SHA256 Hash Check" --text="Computed SHA256 hash:\n$computed_hash\n\nExpected SHA256 hash:\n$expected_hash\n\n$match_message" --width=550



install.sh
#!/bin/bash

# Install script for SHA256 Check - Version 0.1.0

# The directory where the script will be installed
install_dir="$HOME/.local/bin"

# The directory where the Nemo action will be installed
actions_dir="$HOME/.local/share/nemo/actions"

# Function to display a message using zenity
show_message() {
    zenity --info --text="$1" --width=200 --height=100
}

# Function to ask for confirmation
confirm() {
    zenity --question --text="$1" --width=300 --height=100
}

# Ask for confirmation to restart Nemo at the end of the installation
if confirm "The installation will restart Nemo and close all open file browser windows. Do you want to proceed?"; then

    # Creating directories and copying files
    (
    # Create the directories if they do not exist
    mkdir -p "$install_dir" && echo "10" # progress percentage
    mkdir -p "$actions_dir" && echo "20"

    # Copy the script to the install directory and make it executable
    cp sha256_check.sh "$install_dir" && echo "50"
    chmod +x "$install_dir/sha256_check.sh" && echo "70"

    # Use sed to replace the placeholder in the .nemo_action file with the actual install path
    # Then copy the modified .nemo_action file to the actions directory
    sed "s|EXEC_PATH_PLACEHOLDER|$install_dir|g" sha256_check.nemo_action > "$actions_dir/sha256_check.nemo_action" && echo "90"

    # Finish installation
    echo "100"
    ) |
    zenity --progress \
      --title="Installing SHA256 Check" \
      --text="Installing..." \
      --percentage=0 \
      --auto-close \
      --width=300

    # Check if the install directory is in the PATH
    if [[ ":$PATH:" != *":$install_dir:"* ]]; then
        echo "export PATH=\$PATH:$install_dir" >> "$HOME/.profile"
        show_message "Added $install_dir to PATH. Please log out or restart your computer for the changes to take effect."
    fi

    # Restart Nemo
    nemo -q && nemo &

# Give Nemo a little time to close its windows
sleep 2

# Bring the Zenity window back to the front
wmctrl -a "Installing SHA256 Check"

    # Installation complete message
    show_message "Installation complete. Nemo has been restarted for the changes to take effect."
else
    show_message "Installation cancelled by the user."
fi

